lib/tapioca/compilers/dsl_compiler.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+# typed: true
+
+require "tapioca/compilers/dsl/base"
+
+module Tapioca
+  module Compilers
+    class DslCompiler
+      extend T::Sig
+
+      sig { returns(T::Enumerable[Dsl::Base]) }
+      attr_reader :generators
+      sig { returns(T::Array[Module]) }
+      attr_reader :requested_constants
+
+      sig do
+        params(
+          requested_constants: T::Array[Module],
+          requested_generators: T::Array[String]
+        ).void
+      end
+      def initialize(requested_constants:, requested_generators: [])
+        @generators = T.let(
+          gather_generators(requested_generators),
+          T::Enumerable[Dsl::Base]
+        )
+        @requested_constants = requested_constants
+      end
+
+      sig { params(blk: T.proc.params(constant: Module, rbi: String).void).void }
+      def run(&blk)
+        constants_to_process = gather_constants(requested_constants)
+
+        if constants_to_process.empty?
+          $stderr.puts "!!! No classes/modules can be matched for RBI generation."
+          $stderr.puts "!!! Please check that the requested classes/modules include processable DSL methods."
+          exit(1)
+        end
+
+        constants_to_process.each do |constant|
+          rbi = rbi_for_constant(constant)
+          next if rbi.nil?
+
+          blk.call(constant, rbi)
+        end
+      end
+
+      private
+
+      sig { params(requested_generators: T::Array[String]).returns(Proc) }
+      def generator_filter(requested_generators)
+        return proc { true } if requested_generators.empty?
+
+        generators = requested_generators.map(&:downcase)
+
+        proc do |klass|
+          generator = klass.name&.sub(/^Tapioca::Compilers::Dsl::/, '')&.downcase
+          generators.include?(generator)
+        end
+      end
+
+      sig { params(requested_generators: T::Array[String]).returns(T::Enumerable[Dsl::Base]) }
+      def gather_generators(requested_generators)
+        generator_filter = generator_filter(requested_generators)
+
+        Dsl::Base.descendants.select(&generator_filter).map(&:new)
```

Comment:
Only instanciating requested generators by their name will make it harder later if you want to introduce dependencies between them (ie. `GenA`, needs `GenB` to run before it). 

I think it would be better to instantiate all of them, then in a loop ask them if they need to run (and later ask them which other generators need to run before them). We could also imagine passing specific options to each generators and let them parse what they need if we want to alter their behavior from the `cli`.